instrumentation/action_mailer/lib/opentelemetry/instrumentation/action_mailer/instrumentation.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module ActionMailer
+      # The Instrumentation class contains logic to detect and install the ActionMailer instrumentation
+      class Instrumentation < OpenTelemetry::Instrumentation::Base
+        MINIMUM_VERSION = Gem::Version.new('6.1.0')
+        install do |_config|
+          resolve_email_address
+          ecs_mail_convention
+          require_dependencies
+        end
+
+        present do
+          defined?(::ActionMailer)
+        end
+
+        compatible do
+          gem_version >= MINIMUM_VERSION
+        end
+
+        option :disallowed_notification_payload_keys, default: [], validate: :array
+        option :notification_payload_transform,       default: nil, validate: :callable
+        option :email_address,                        default: :omit, validate: %I[omit include]
+
+        private
+
+        def gem_version
+          ::ActionMailer.version
+        end
+
+        def resolve_email_address
+          return unless ActionMailer::Instrumentation.instance.config[:email_address] == :omit
+
+          ActionMailer::Instrumentation.instance.config[:disallowed_notification_payload_keys] += ['email.to.address', 'email.from.address', 'email.cc.address', 'email.bcc.address']
+        end
+
+        def ecs_mail_convention
+          if ActionMailer::Instrumentation.instance.config[:notification_payload_transform].nil?
+            transform_attributes = ->(payload) { transform_payload(payload) }
+          else
+            original_callable = ActionMailer::Instrumentation.instance.config[:notification_payload_transform]
+            transform_attributes = lambda do |payload|
+              new_payload = transform_payload(payload)
+              user_payload = original_callable.call(new_payload)
+              if user_payload.instance_of?(Hash)
```

Comment:
what are we protecting from here? The user passing in an invalid transformer that does not return a hash?

Should this actually be in the `ActiveSupport::Notifications` instrumentation instead?

Reply:
> The user passing in an invalid transformer that does not return a hash?  

To ensure user's callable return hash, otherwise the active_support part will report error.

> Should this actually be in the ActiveSupport::Notifications instrumentation instead?

It could be.

Also, just want to pitch the idea of restriction of mutating original payload here:

When user provide callable function for this instrumentation, they can only play with the transformed payload that is generated from `transform_payload()`, not the original payload. This avoid user mutating original payload in this instrumentation. 

They can still mutate the payload if they subscribe the same subscription pattern (i.e. `deliver.action_mailer`) but that's up to user.